keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,249 @@
+  information is hard-coded, like in this [list][].
+* creates `migration` CRs (see the [API section][] for the schema) for all
+  remaining resources. The `ownerReferences` of the `migration` objects are set
+  to the **kube-migrator controller** `deployment`. Thus, the old `migration`s
+  are deleted with the old `deployment` in the first step.
+* uses the same kubeconfig file as **kubectl** does.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR, adds them to a [workqueue][].
+* syncs one `migration` at a time to avoid overloading the apiserver,
+  * if `migration.status.completed == false`, it creates a **migration worker**
+    `job` to migrate the resource type. The `ownerReference` of the `job` is set
+    to the `migration`.
+  * waits until the **migration worker** `job` finishes, updates
+    `migration.status.completed` to `true`, and deletes the `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in the owner
+`migration.status.continueToken`. With the inconsistent continue token
+introduced in [#67284][], the **migrator worker** does not need to worry about
+expired continue token.
+
+[list]:https://github.com/openshift/origin/blob/2a8633598ef0dcfa4589d1e9e944447373ac00d7/pkg/oc/cli/admin/migrate/storage/storage.go#L120-L184
+[workqueue]:https://github.com/kubernetes/client-go/tree/master/util/workqueue
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#alpha-api
+
+The cluster admin can run the `kubectl get migrations -o
+jsonpath="{.status.completed}"` to see if all migrations have completed.
+
+We will support a `kube-migrator migrate --resources=<list of resources>`
+command. It (re)creates `migration` CRs for the resources to trigger their
+migration. This command facilitates manually migrating custom resources and
+aggregated resources.
+
+### Alpha API
+
+We introduce the `migration` API to record the intention and the progress of a
+migration. The API will be a CRD defined in the `migration.k8s.io` group.
+
+Read the [workflow section][] to understand how the API is used.
+
+```golang
+type StorageVersionMigration struct {
+  metav1.TypeMeta
+  // metadata.name should be "<resource>.<group>" of the resource being
+  // migrated.
+  // Because resouces can be served at different groups, in the future version,
+  // metadata.name can be hash(etcd instance, etcd path prefix).
+  metav1.ObjectMeta
+  Spec StorageVersionMigrationSpec
+  Status StorageVersionMigrationStatus
+}
+
+type StorageVersionMigrationSpec {
+  // Resource is the resource that is being migrated. The migrator sends
+  // requests to the endpoint serving the Resource.
+  // Immutable.
```

Comment:
if this is immutable, how do I update a field in here to indicate that the generation needs to be bumped?  The reason field made such a thing a possible and allowed me to key a request to an external system with an opaque key.

Reply:
I'm thinking in the future we will add a mutable field to this spec, like `targetStorageVersion(Hash)`, then the spec will be changing.

My concern with a `Reason string` field is it's too loose. User might edit the `Reason` and inadvertently triggered a new round of migration.

Can the external system to be triggered by the `migration.status.conditions == proposed`?

Reply:
> Can the external system to be triggered by the `migration.status.conditions == proposed`?

The external system drives this migration.  It needs the generation of this object to be incremented to do that.

Reply:
Why does the generation need to change if you are deleting and recreating to request a new migration?

Reply:
I think three of us had three ideas on how to drive the migration control loop. I updated the KEP to follow Daniel's idea.

The original three ideas are:

I wanted to maintain one `migration` CR for one resource type, and the control loop would start the migration if `migration.status==nil`. User needs to delete and create `migration` to request for another round of migration.

I think deads2k's model was also one `migration` CR for one resource type, but using the `migration.metadata.generation != observedGeneration` to drive the control loop to do migration.

In lavalamp's model, users *can* create multiple `migration` CRs for a single resource type. Each CR will cause the control loop to do the migration once. This results in wasteful migration, but doesn't affect correctness. This state is recoverable by deleting redundant `migration`s. And we won't reach this state unless users are manually creating duplicate `migration` CRs. I updated the KEP to lavalamp's idea because it's the simplest and low risk.

Reply:
> I think three of us had three ideas on how to drive the migration control loop. I updated the KEP to follow Daniel's idea.
> 
> The original three ideas are:
> 
> I wanted to maintain one `migration` CR for one resource type, and the control loop would start the migration if `migration.status==nil`. User needs to delete and create `migration` to request for another round of migration.
> 
> I think deads2k's model was also one `migration` CR for one resource type, but using the `migration.metadata.generation != observedGeneration` to drive the control loop to do migration.
> 
> In lavalamp's model, users _can_ create multiple `migration` CRs for a single resource type. Each CR will cause the control loop to do the migration once. This results in wasteful migration, but doesn't affect correctness. This state is recoverable by deleting redundant `migration`s. And we won't reach this state unless users are manually creating duplicate `migration` CRs. I updated the KEP to lavalamp's idea because it's the simplest and low risk.

I think that I could use the model multiple migration CRs for a single resource type.  It's a more batch-like way of looking at the problem, but I think it can work.  Proof of the pudding is in the tasting - I'm willing to taste it and see.